internal/query/relations.go

```diff
@@ -267,16 +269,166 @@ func (e *engine) CreateRelationships(ctx context.Context, rels []types.Relations
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.Role{}, err
+	}
+
+	if err = dbTx.Commit(); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
 		return types.Role{}, err
 	}
 
+	role.Creator = dbTx.Record.CreatorID
+	role.ResourceID = dbTx.Record.ResourceID
+	role.CreatedAt = dbTx.Record.CreatedAt
+	role.UpdatedAt = dbTx.Record.UpdatedAt
+
+	return role, nil
+}
+
+// actionsDiff determines which actions needs to be added and removed.
+// If no new actions are provided it is assumed no changes are requested.
+func actionsDiff(oldActions, newActions []string) ([]string, []string) {
+	if len(newActions) == 0 {
+		return nil, nil
+	}
+
+	old := make(map[string]bool, len(oldActions))
+	new := make(map[string]bool, len(newActions))
+
+	var add, rem []string
+
+	for _, action := range oldActions {
+		old[action] = true
+	}
+
+	for _, action := range newActions {
+		new[action] = true
+
+		// If the new action is not in the old actions, then we need to add the action.
+		if !old[action] {
+			add = append(add, action)
+		}
+	}
+
+	for _, action := range oldActions {
+		// If the old action is not in the new actions, then we need to remove it.
+		if !new[action] {
+			rem = append(rem, action)
+		}
+	}
+
+	return add, rem
+}
+
+// UpdateRole allows for updating an existing role with a new name and new actions.
+// If new name is empty, no change is made.
+// If new actions is an empty slice, no change is made.
+func (e *engine) UpdateRole(ctx context.Context, actor, roleResource types.Resource, newName string, newActions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.UpdateRole")
+
+	defer span.End()
+
+	role, err := e.GetRole(ctx, roleResource)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	newName = strings.TrimSpace(newName)
+
+	addActions, remActions := actionsDiff(role.Actions, newActions)
+
+	// If no changes, return existing role with no changes.
+	if role.Name == newName && len(addActions) == 0 && len(remActions) == 0 {
+		return role, nil
+	}
+
+	resourceID := role.ResourceID
+
+	// If the resource id is not found in the permissions database, then we must locate it in the spicedb database.
+	if resourceID == gidx.NullPrefixedID {
+		resource, err := e.GetRoleResource(ctx, roleResource)
+		if err != nil {
+			return types.Role{}, fmt.Errorf("failed to locate roles associated resource: %s: %w", roleResource.ID.String(), err)
+		}
+
+		resourceID = resource.ID
+	}
+
+	resource, err := e.NewResourceFromID(resourceID)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	var (
+		dbTx  database.TxRole
+		dbErr error
+	)
+
+	// If new name has changed, commit change to permissions database.
+	if newName != "" && role.Name != newName {
+		dbTx, dbErr = e.db.UpdateRoleTransaction(ctx, actor.ID, role.ID, newName, resourceID)
+		if dbErr != nil {
+			return types.Role{}, dbErr
+		}
+
+		defer dbTx.Rollback() //nolint:errcheck // error is logged in function
+	}
```

Comment:
We always want to update the role somehow, even if it's just to change the updated timestamp.